### ISSUE TITLE
dropping many (all?) foreign keys after update from 1.x to 2.x

### ISSUE THREAD

**Issue by obelixA** [aspnet/EntityFrameworkCore#10617]
Hi
I just tried to update all my packages via nuget.
compiling works fine, but adding a migration (I made no changes to the source - just tried it out to see if something changes AND IT DOES!)
I can see many of these drops
```
migrationBuilder.DropForeignKey(
    name: "FK_AZAbschnitt_AZTage_FK_AZTag",
    table: "AZAbschnitt");
```
only to see a creation of the keys like
```
migrationBuilder.AlterColumn<int>(
                name: "FK_AZTag",
                table: "AZAbschnitt",
                nullable: false,
                oldClrType: typeof(int),
                oldNullable: true);
```
as far as I can see the only difference is, that now the entries are not nullable.
BUT
as I use sqlite the drops are not working. And there is no need to change anything with the source code not changed.
What can I add/change/do that my existing databases are still working with the new sqlite-/ef core libraries?

any hints are welcome :)

**Comment by obelixA** [comment]
searching for a solution that helps me out, I found that this has been discussed here: [here](https://github.com/aspnet/EntityFrameworkCore/issues/3751)
and should have been fixed, but I still get this.
It is caused by `onDelete: ReferentialAction.Cascade` which is added automatically (sometimes). Most of my migrations work as they are, but with some the onDelete behavior is changed from "restrict" to "cascade". This leads to dropping and recreating all foreign keys. of the context.
my next try is manually changing "restrict" to "cascade" and hope that everything works again...

If someone has hints for me how to change the migrations securely, please tell me...

**Comment by ajcvickers** [comment]
@obelixA Could you please post a runnable project/solution or full code listing that reproduces what you are seeing? Also, what version of EF Core was use to generate the migrations--were they all created with 1.x, or were any of them created with prerelease versions?

**Comment by obelixA** [comment]
no prerelease version were involved.
here come the initial migration and the corresponding snapshot, followed by the classes that generated them.
`
using System;
using System.Collections.Generic;
using Microsoft.EntityFrameworkCore.Migrations;

namespace ActiveCashSystem.MigrationsacsFactura
{
    public partial class acsFactura_Initial : Migration
    {
        protected override void Up(MigrationBuilder migrationBuilder)
        {
            migrationBuilder.CreateTable(
                name: "factura",
                columns: table => new
                {
                    LfdNr = table.Column<int>(nullable: false)
                        .Annotation("Sqlite:Autoincrement", true),
                    ArtikelPreiseNetto = table.Column<bool>(nullable: false),
                    Bemerkung = table.Column<string>(nullable: true),
                    Bezahlt = table.Column<bool>(nullable: false),
                    DocumentName = table.Column<string>(nullable: true),
                    ErstalltAm = table.Column<DateTime>(nullable: false),
                    ErstelltAusLfdNr = table.Column<int>(nullable: false),
                    Gesamt = table.Column<decimal>(nullable: false),
                    GesamtRabatt = table.Column<decimal>(nullable: false),
                    ID = table.Column<Guid>(nullable: false),
                    KundeLiefer = table.Column<int>(nullable: false),
                    KundeRechnung = table.Column<int>(nullable: false),
                    Kundenreferenz = table.Column<string>(nullable: true),
                    Leistungsdatum = table.Column<DateTime>(nullable: false),
                    LieferAdresse = table.Column<string>(nullable: true),
                    Nummer = table.Column<string>(nullable: true),
                    RechnungsAdresse = table.Column<string>(nullable: true),
                    SteuerVorlageID_as_string = table.Column<string>(nullable: true),
                    VersandNetto = table.Column<decimal>(nullable: false),
                    VersandSteuer = table.Column<decimal>(nullable: false),
                    fType = table.Column<int>(nullable: false)
                },
                constraints: table =>
                {
                    table.PrimaryKey("PK_factura", x => x.LfdNr);
                });

            migrationBuilder.CreateTable(
                name: "nextNumbers",
                columns: table => new
                {
                    LfdNr = table.Column<int>(nullable: false)
                        .Annotation("Sqlite:Autoincrement", true),
                    ID = table.Column<Guid>(nullable: false),
                    NextNumber = table.Column<int>(nullable: false),
                    fType = table.Column<int>(nullable: false)
                },
                constraints: table =>
                {
                    table.PrimaryKey("PK_nextNumbers", x => x.LfdNr);
                });

            migrationBuilder.CreateTable(
                name: "acsFacturaLine",
                columns: table => new
                {
                    LfdNr = table.Column<int>(nullable: false)
                        .Annotation("Sqlite:Autoincrement", true),
                    AnzRueck = table.Column<int>(nullable: false),
                    Anzahl = table.Column<int>(nullable: false),
                    ArtikelNr = table.Column<string>(nullable: true),
                    Beschreibung = table.Column<string>(nullable: true),
                    Bezeichnung = table.Column<string>(nullable: true),
                    EinzelNettoPreis = table.Column<decimal>(nullable: false),
                    EinzelPreis = table.Column<decimal>(nullable: false),
                    FK_FacturaBase = table.Column<int>(nullable: true),
                    ID = table.Column<Guid>(nullable: false),
                    NettoPreis = table.Column<decimal>(nullable: false),
                    Preis = table.Column<decimal>(nullable: false),
                    RabattInPrz = table.Column<decimal>(nullable: false),
                    Steuer = table.Column<decimal>(nullable: false)
                },
                constraints: table =>
                {
                    table.PrimaryKey("PK_acsFacturaLine", x => x.LfdNr);
                    table.ForeignKey(
                        name: "FK_acsFacturaLine_factura_FK_FacturaBase",
                        column: x => x.FK_FacturaBase,
                        principalTable: "factura",
                        principalColumn: "LfdNr",
                        onDelete: ReferentialAction.Restrict);
                });

            migrationBuilder.CreateTable(
                name: "acsFacturaPayment",
                columns: table => new
                {
                    LfdNr = table.Column<int>(nullable: false)
                        .Annotation("Sqlite:Autoincrement", true),
                    Betrag = table.Column<decimal>(nullable: false),
                    Datum = table.Column<DateTime>(nullable: false),
                    FK_FacturaBase = table.Column<int>(nullable: true),
                    ID = table.Column<Guid>(nullable: false),
                    Kommentar = table.Column<string>(nullable: true)
                },
                constraints: table =>
                {
                    table.PrimaryKey("PK_acsFacturaPayment", x => x.LfdNr);
                    table.ForeignKey(
                        name: "FK_acsFacturaPayment_factura_FK_FacturaBase",
                        column: x => x.FK_FacturaBase,
                        principalTable: "factura",
                        principalColumn: "LfdNr",
                        onDelete: ReferentialAction.Restrict);
                });

            migrationBuilder.CreateIndex(
                name: "IX_acsFacturaLine_FK_FacturaBase",
                table: "acsFacturaLine",
                column: "FK_FacturaBase");

            migrationBuilder.CreateIndex(
                name: "IX_acsFacturaPayment_FK_FacturaBase",
                table: "acsFacturaPayment",
                column: "FK_FacturaBase");
        }

        protected override void Down(MigrationBuilder migrationBuilder)
        {
            migrationBuilder.DropTable(
                name: "acsFacturaLine");

            migrationBuilder.DropTable(
                name: "nextNumbers");

            migrationBuilder.DropTable(
                name: "acsFacturaPayment");

            migrationBuilder.DropTable(
                name: "factura");
        }
    }
}
`
`using System;
using Microsoft.EntityFrameworkCore;
using Microsoft.EntityFrameworkCore.Infrastructure;
using Microsoft.EntityFrameworkCore.Metadata;
using Microsoft.EntityFrameworkCore.Migrations;
using ActiveCashSystem;

namespace ActiveCashSystem.MigrationsacsFactura
{
    [DbContext(typeof(acsFacturaContext))]
    partial class acsFacturaContextModelSnapshot : ModelSnapshot
    {
        protected override void BuildModel(ModelBuilder modelBuilder)
        {
            modelBuilder
                .HasAnnotation("ProductVersion", "1.1.0-rtm-22752");

            modelBuilder.Entity("ActiveCashSystem.acsFacturaBase", b =>
                {
                    b.Property<int>("LfdNr")
                        .ValueGeneratedOnAdd();

                    b.Property<bool>("ArtikelPreiseNetto");

                    b.Property<string>("Bemerkung");

                    b.Property<bool>("Bezahlt");

                    b.Property<string>("DocumentName");

                    b.Property<DateTime>("ErstalltAm");

                    b.Property<int>("ErstelltAusLfdNr");

                    b.Property<decimal>("Gesamt");

                    b.Property<decimal>("GesamtRabatt");

                    b.Property<Guid>("ID");

                    b.Property<int>("KundeLiefer");

                    b.Property<int>("KundeRechnung");

                    b.Property<string>("Kundenreferenz");

                    b.Property<DateTime>("Leistungsdatum");

                    b.Property<string>("LieferAdresse");

                    b.Property<string>("Nummer");

                    b.Property<string>("RechnungsAdresse");

                    b.Property<string>("SteuerVorlageID_as_string");

                    b.Property<decimal>("VersandNetto");

                    b.Property<decimal>("VersandSteuer");

                    b.Property<int>("fType");

                    b.HasKey("LfdNr");

                    b.ToTable("factura");
                });

            modelBuilder.Entity("ActiveCashSystem.acsFacturaLine", b =>
                {
                    b.Property<int>("LfdNr")
                        .ValueGeneratedOnAdd();

                    b.Property<int>("AnzRueck");

                    b.Property<int>("Anzahl");

                    b.Property<string>("ArtikelNr");

                    b.Property<string>("Beschreibung");

                    b.Property<string>("Bezeichnung");

                    b.Property<decimal>("EinzelNettoPreis");

                    b.Property<decimal>("EinzelPreis");

                    b.Property<int?>("FK_FacturaBase");

                    b.Property<Guid>("ID");

                    b.Property<decimal>("NettoPreis");

                    b.Property<decimal>("Preis");

                    b.Property<decimal>("RabattInPrz");

                    b.Property<decimal>("Steuer");

                    b.HasKey("LfdNr");

                    b.HasIndex("FK_FacturaBase");

                    b.ToTable("acsFacturaLine");
                });

            modelBuilder.Entity("ActiveCashSystem.acsFacturaNextNumbers", b =>
                {
                    b.Property<int>("LfdNr")
                        .ValueGeneratedOnAdd();

                    b.Property<Guid>("ID");

                    b.Property<int>("NextNumber");

                    b.Property<int>("fType");

                    b.HasKey("LfdNr");

                    b.ToTable("nextNumbers");
                });

            modelBuilder.Entity("ActiveCashSystem.acsFacturaPayment", b =>
                {
                    b.Property<int>("LfdNr")
                        .ValueGeneratedOnAdd();

                    b.Property<decimal>("Betrag");

                    b.Property<DateTime>("Datum");

                    b.Property<int?>("FK_FacturaBase");

                    b.Property<Guid>("ID");

                    b.Property<string>("Kommentar");

                    b.HasKey("LfdNr");

                    b.HasIndex("FK_FacturaBase");

                    b.ToTable("acsFacturaPayment");
                });

            modelBuilder.Entity("ActiveCashSystem.acsFacturaLine", b =>
                {
                    b.HasOne("ActiveCashSystem.acsFacturaBase", "acsF")
                        .WithMany("Lines")
                        .HasForeignKey("FK_FacturaBase");
                });

            modelBuilder.Entity("ActiveCashSystem.acsFacturaPayment", b =>
                {
                    b.HasOne("ActiveCashSystem.acsFacturaBase", "acsF")
                        .WithMany("Zahlungen")
                        .HasForeignKey("FK_FacturaBase");
                });
        }
    }
}
`

`public class acsFacturaNextNumbers : AcsDBBaseType {
        public acsFacturaType fType { get; set; }
        public int NextNumber { get; set; }
    }

    public enum acsFacturaType {
        acsAngebot = 0,
        acsAuftrag = 1,
        acsLieferschein = 2,
        acsRechnung = 3,
        acsGutschrift = 4,
        acsMahnung = 5,
        acsBestellung = 6
    }

    public class acsFacturaPayment : AcsDBBaseType {
        public DateTime Datum { get; set; } = DateTime.Now;
        public Decimal Betrag { get; set; } = 0;
        public string Kommentar { get; set; } = String.Empty;

        [ForeignKey( "FK_FacturaBase" )]
        public acsFacturaBase acsF { get; set; }
    }

    public class acsFacturaLine : AcsDBBaseType, ICloneable {
        public int Anzahl { get; set; } = 1;
        public int AnzRueck { get; set; } = 0;
        public string ArtikelNr { get; set; } = String.Empty;
        public string Bezeichnung { get; set; } = String.Empty;
        public string Beschreibung { get; set; } = String.Empty;
        public Decimal Steuer { get; set; } = 19;
        private Decimal einzelPreis = 0;
        public Decimal EinzelPreis {
            get { return einzelPreis; }
            set {
                einzelPreis = value;
                einzelNettoPreis = Math.Round( einzelPreis / (Steuer / 100 + 1), 2 );
            }
        }
        private Decimal einzelNettoPreis = 0;
        public Decimal EinzelNettoPreis {
            get { return einzelNettoPreis; }
            set {
                einzelNettoPreis = value;
                einzelPreis = Math.Round( einzelNettoPreis + einzelNettoPreis * Steuer / 100, 2 );
            }
        }
        public Decimal RabattInPrz { get; set; } = 0;
        private Decimal preis = 0;
        public Decimal Preis {
            get { return preis; }
            set {
                preis = value;
                nettoPreis = Math.Round( preis / (Steuer / 100 +1), 2 );
            }
        }
        private Decimal nettoPreis = 0;
        public Decimal NettoPreis {
            get { return nettoPreis; }
            set {
                nettoPreis = value;
                preis = Math.Round( nettoPreis + nettoPreis * Steuer / 100, 2 );
            }
        }

        [ForeignKey( "FK_FacturaBase" )]
        public acsFacturaBase acsF { get; set; }

        public object Clone() {
            acsFacturaLine ln = new acsFacturaLine();
            ln.Anzahl = Anzahl;
            ln.AnzRueck = AnzRueck;
            ln.ArtikelNr = ArtikelNr;
            ln.Bezeichnung = Bezeichnung;
            ln.Beschreibung = Beschreibung;
            ln.Steuer = Steuer;
            ln.EinzelPreis = EinzelPreis;
            ln.RabattInPrz = RabattInPrz;
            ln.Preis = Preis;
            return ln;
        }
    }

    public class acsFacturaBase : AcsDBBaseType, ICloneable {
        public acsFacturaType fType { get; set; } = acsFacturaType.acsAngebot;
        public string Nummer { get; set; } = String.Empty;
        public DateTime ErstalltAm { get; set; } = DateTime.Now;
        public DateTime Leistungsdatum { get; set; } = DateTime.Now;
        public string DocumentName { get; set; } = String.Empty;
        [NotMapped]
        public string DocumentFullName {
            get {
                return String.Format( "0:000000", int.Parse( Nummer ) +" " + DocumentName );
            }
        }
        public Decimal Gesamt { get; set; } = 0;
        private Guid steuerVorlageID = Guid.Empty;
        [NotMapped]
        public Guid SteuerVorlageID {
            get { return steuerVorlageID; }
            set {
                steuerVorlageID = value;
            }
        }
        public string SteuerVorlageID_as_string {
            get { return steuerVorlageID.ToString(); }
            set { steuerVorlageID = Guid.Parse( value ); }
        }
        public string Kundenreferenz { get; set; } = String.Empty;
        [ForeignKey( "FK_Kunde_LfdNr" )]
        public int KundeRechnung { get; set; }
        public string RechnungsAdresse { get; set; } = String.Empty;
        [ForeignKey( "FK_Kunde_LfdNr" )]
        public int KundeLiefer { get; set; }
        public string LieferAdresse { get; set; } = String.Empty;

        public Decimal GesamtRabatt { get; set; } = 0;
        public string Bemerkung { get; set; } = String.Empty;
        public Decimal VersandNetto { get; set; } = 0;
        public Decimal VersandSteuer { get; set; } = 19;
        public bool Bezahlt { get; set; } = false;
        public bool ArtikelPreiseNetto { get; set; } = true;
        public ICollection<acsFacturaPayment> Zahlungen { get; set; } = new List<acsFacturaPayment>();
        public ICollection<acsFacturaLine> Lines { get; set; } = new List<acsFacturaLine>();
        public int ErstelltAusLfdNr { get; set; } = 0;

        public object Clone() {
            acsFacturaBase data = new acsFacturaBase();
            data.fType = fType;
            data.Nummer = String.Empty;
            data.ErstalltAm = ErstalltAm;
            data.Leistungsdatum = Leistungsdatum;
            data.DocumentName = DocumentName;
            data.Gesamt = Gesamt;
            data.steuerVorlageID = steuerVorlageID;
            data.Kundenreferenz = Kundenreferenz;
            data.KundeRechnung = KundeRechnung;
            data.KundeLiefer = KundeLiefer;
            data.RechnungsAdresse = RechnungsAdresse;
            data.LieferAdresse = LieferAdresse;
            data.GesamtRabatt = GesamtRabatt;
            data.Bemerkung = Bemerkung;
            data.VersandNetto = VersandNetto;
            data.VersandSteuer = VersandSteuer;
            data.ArtikelPreiseNetto = ArtikelPreiseNetto;
            foreach( acsFacturaLine ln in Lines )
                data.Lines.Add( (acsFacturaLine)ln.Clone() );
            return data;
        }
    }`

**Comment by bricelam** [comment]
This looks like a breaking change in model building in 2.0.

``` csharp


    class Line
    {
        [Key]
        public int Key { get; set; }

        [ForeignKey("FK_Bill")]
        public Bill Bill { get; set; }
    }

    class Bill
    {
        [Key]
        public int Key { get; set; }

        public ICollection<Line> Lines { get; set; }
    }
```

In 1.1, `FK_Bill` is nullable. In 2.0, it's not.

Interestingly, in 2.0 when the keys are named `Id` or they are different from each other (e.g. `LineKey` and `BillKey`) it becomes nullable again.

@obelixA you can work around the change by adding the following to `OnModelCreating()`.

``` csharp
modelBuilder.Entity<acsFacturaPayment>().Property("FK_FacturaBase").IsRequired(false);
modelBuilder.Entity<acsFacturaLine>().Property("FK_FacturaBase").IsRequired(false);
```

**Comment by AndriySvyryd** [comment]
Fixed in 18fb0ff310c0f9af6163c233ab187646863a3c6e

**Comment by obelixA** [comment]
Trying to make changes on my data model lead me to the following problems:

The property 'FK_AZTag' on entity type 'AZAbschnitt' cannot be marked as nullable/optional because the type of the property is 'int' which is not a nullable type. Any property can be marked as non-nullable/required, but only properties of nullable types and which are not part of primary key can be marked as nullable/optional.

this was the advice [https://github.com/aspnet/EntityFrameworkCore/issues/10617#issuecomment-356099421] some lines above by @bricelam 

commenting out the lines in the source leads to dropping and recreating the keys.

So are there any new rules how to fix this? 
btw: I already tried to manually change the fields from <int?> to <int> and from restrict to cascade and many other unsuccessful tries... I have no idea how I can continue...
Any advice is appreciated...

**Comment by ajcvickers** [comment]
@obelixA This issue is fixed. Please file a new issue including a small, runnable project/solution or complete code listing that demonstrates the behavior you are seeing.

**Comment by obelixA** [comment]
the code is exactly the one provided above. My problem is, that the hint how to fix the changed from ef core 1 to ef core 2 do not work anymore. Now I get exactly the given exception and I do have no chance to surround it. neither by changing int to int? or vice versa nor by trying to manually change items within the auto-generated code. 
Updating to the latest (ef core 3 preview) libs leads to unrunnable code and thrown exceptions. 
Going back to version 2.1.0 of all ef core libs (and related) makes my program run again, but no chance to extend the data types for the DB.
If the above provided code does not satisfy the needs for providing hints for a new try, I will try to extract some lines of code from the project for you. But that is a difficult issue since there are 86 projects in the solution...

**Comment by ajcvickers** [comment]
@obelixA The fix for this issue was released in the EF Core 2.1 preview, which means it is also fixed in EF Core 2.1 RTM. Are you saying that EF Core 2.1 release didn't fix the issue for you? If so, then please open a new issue with a runnable repro, because either what you are hitting is a different case that isn't fix, or the fix we made was missing something. In either case, we'll need to be able to reproduce exactly what you are hitting.
